torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -269,14 +269,36 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  rpcRunning_.store(true);
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
   futureTimeoutThread_ =
       std::thread(&ProcessGroupAgent::pollTimedOutRPCs, this);
 }
 
+void ProcessGroupAgent::shutdown() {
+  if (!rpcRunning_.load()) {
```

Comment:
You're using an atomic, but the check isn't atomic. Multiple threads may call `shutdown` and both of them will execute this function, if they access the atomic's value before one of them has a chance to update it. To make this correct while retaining the atomic (instead of using a mutex every time you access it), you'll have to do do an atomic exchange.

For example:
```c++
// If the value of rpcRunning_ was false already, don't do anything.
if (!rpcRunning_.exchange(false)) {
  return;
}
// Only one thread will have seen the edge from true -> false
```

That said, if you're looking for this behavior AND the guarantee that the agent has shut down when this function returns, check out `std::call_once`. Then, the second thread to call `shutdown` will only return if the first thread has also returned.

Reply:
Thanks for this catch. I think the goal is to prevent shutting down twice, as then we'd have issues where we call `join` on threads that have already been `join`ed, for example. I'm also not sure where we could have multiple threads call this shutdown function, as it's only exposed in the python code, so if the user launches multiple threads that all call shutdown, this could be an issue, though that is an unexpected use case.